### PR TITLE
render breadcrumbs on page

### LINKED ISSUE #128: render breadcrumbs on pageheader component
To improve navigation, we might want to add breadcrumbs to every page. The way to go here is to make use of Pattenfly's [breadcrumb component](https://www.patternfly.org/2020.06/documentation/react/components/breadcrumb).

### PR DESCRIPTION
**Pull request by aquibbaig**
Fixes #128

### CODE REVIEW

**Comment by anishaswain** [comment]
@aquibbaig is this PR ready for review yet?

**Comment by aquibbaig** [comment]
@anishaswain Actually I am waiting for Mary for some designs regarding the placement of breadcrumbs w.r.t header mast. We're almost there implementation wise

**Comment by anishaswain** [comment]
@aquibbaig @gurbirkalsi while going through the PR, I noticed that two words like `Comparison Select` or `Run Comparison` looks a bit funky on the breadcrumbs. Will it be a good idea to combine both words with a `-` or CamelCase instead?
Screenshot 2021-05-05 at 4 35 27 PM

**Comment by aquibbaig** [comment]
image
*went ahead with the second design

File: src/common/menu.js [hunk 1]
```diff
@@ -37,19 +37,35 @@ const menuData = [
...
+    // For all non root paths,
+    // append the parent name with
+    // current name.
+    if (level !== 1) {
+      name = `${parentName}/${name}`;
+    }
+    let result = {};
     if (item.routes) {
-      result.routes = formatter(item.routes, `${parentPath}${item.path}`, item.authority);
+      result = {
```
Comment (gurbirkalsi) [inline src/common/menu.js]: We can avoid duplicate assignment to `result` in the if else block by moving it outside of the block (DRY principle).

File: src/components/LoginHint/index.js [hunk 2]
```diff
@@ -1,7 +1,9 @@
 import React, { Component } from 'react';
 import { routerRedux } from 'dva/router';
 import { connect } from 'dva';
-import { Title, Button } from '@patternfly/react-core';
+import { Title, Button, Banner, Flex, FlexItem } from '@patternfly/react-core';
```
Comment (gurbirkalsi) [inline src/components/LoginHint/index.js:4]: These changes seem unrelated to the PR. Did you mean to include changes to update the login hint UI/UX?

File: src/layouts/BasicLayout.js [hunk 3]
```diff
@@ -87,27 +82,52 @@ class BasicLayout extends React.PureComponent {
...
-    return {
-      location,
-      breadcrumbNameMap: this.breadcrumbNameMap,
-      routes: route.routes,
-    };
+  componentDidMount() {
+    const {
+      location: { pathname },
+    } = this.props;
+    let exactPath = {};
```
Comment (gurbirkalsi) [inline src/layouts/BasicLayout.js]: We should consider creating a function for the logic on lines 94 - 102 to avoid repeating duplicate functionality (DRY principle).

File: src/utils/utils.js [hunk 4]
```diff
@@ -104,3 +104,38 @@ export const isLoggedInUser = username => {
   }
   return false;
 };
+
+export const isRootPath = pathname => {
+  if (pathname === '/' || pathname === '/search' || pathname === '/sessions') {
```
Comment (gurbirkalsi) [inline src/utils/utils.js]: We should consider importing the `menuData` object in `menu.js` instead of hardcoding the root path entrypoints for the dashboard. We can do this by making `menuData` exportable and iterating over each object in the array.

File: src/components/LoginHint/index.js (hunk 2 above)
Comment (aquibbaig) [reply src/components/LoginHint/index.js:4]: Yeah, so we had a discussion in the design sync to add the login banner rather for a hint

File: src/components/Breadcrumb/index.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    return [...pathTo(route.parent), route];
+  };
+
+  const navigateToCrumb = path => {
+    dispatch(routerRedux.push(path));
+  };
+
+  return (
```
Comment (anishaswain) [inline src/components/Breadcrumb/index.js:20]: @aquibbaig trying using the below code for rendering arrow icon.
```
<Breadcrumb>
      {pathTo(context).map((crumb, index, breadcrumbs) => (
            <BreadcrumbItem
            className={index < breadcrumbs.length - 1?styles.active:''}
            key={crumb.path}
            onClick={() => navigateToCrumb(crumb.path)}
            isActive={index === breadcrumbs.length - 1?true:false}
          >
            {crumb.name}
          </BreadcrumbItem>
        )
      )}
</Breadcrumb>
 ```

**Review by anishaswain — CHANGES_REQUESTED** [review]

File: src/components/Breadcrumb/index.js (hunk 5 above)
Comment (aquibbaig) [reply src/components/Breadcrumb/index.js:20]: That works, thanks @anishaswain

**Comment by aquibbaig** [comment]
@anishaswain mergeable?

**Review by anishaswain — COMMENTED** [review]
@aquibbaig LGTM! @aquibbaig could you please squash the commits before we merge.

File: src/common/menu.js [hunk 6]
```diff
@@ -37,23 +37,37 @@ const menuData = [
...
+      ...(withParent.routes && {
+        routes: setupParents(withParent.routes, withParent),
+      }),
+    };
+  });
+
+const flattenRoutes = routes =>
+  routes.map(route => [route.routes ? flattenRoutes(route.routes) : [], route]).flat(Infinity);
+
+const getMenuData = () => flattenRoutes(setupParents(buildPaths(menuData)));
```
Comment (portante) [inline src/common/menu.js:71]: Why do we still need this?

**Review by portante — CHANGES_REQUESTED** [review]

File: src/common/menu.js (hunk 6 above)
Comment (aquibbaig) [reply src/common/menu.js:71]: Good question @portante!
umijs requires you to define your routing configuration in a config file. We do not instantiate any top-level declaration w.r.t routes as we would have done with something like  [react-router-dom](https://reactrouter.com/web/guides/quick-start).
So, we generate the breadcrumbs array (required by the patternfly breadcrumbs component) by doing some post-processing based on menudata (used for routing) and the current location of the user (passed in by the page). Earlier there was a lot of unnecessary post-processing which I have removed in this PR.

**Review by portante — APPROVED** [review]

**Review by anishaswain — DISMISSED** [review]
LGTM! Just need to resolve the conflicts. @aquibbaig could you please resolve the conflicts.

**Review by portante — CHANGES_REQUESTED** [review]
@aquibbaig, looks like this PR needs a rebase.

**Review by portante — DISMISSED** [review]
